### ISSUE TITLE
Particle reflection examples

### ISSUE THREAD

**Issue by mkeilman** [radiasoft/rswarp#19]
I have done a lot on the sirepo interface - I don't want to make an official pull request until the reflection code gets merged (I see there is a [PR](https://github.com/radiasoft/rswarp/pull/16) for it).  At this point examples would be great.  In the meantime I have a few questions:

1. Which reports are affected by reflections?  Just the particle report or the various field reports as well?
2. Does an object have to be a conductor in the physics sense (not the warp sense where everything seems to be called a "conductor") to support reflection?  For the UI that means only showing the reflection toggle if the conductor toggle is on vs. always showing it
3. Does the voltage of a reflecting object have to be non-zero?
4. Are both 2d and 3d supported?  What about stl-based conductors?
5. Can the anode and cathode have reflection?
6. Should the UI ask which ```refscheme``` (random/uniform) to use, and does that only pertain to non-zero diffuse probability?

Thanks

**Comment by ncook882** [comment]
Hi @mkeilman. Thanks for the summary and questions. I will attempt to answer them below:

1. The two reports affected by reflections are the Particle Trace report and the Current report. 

    Because the particle reflector class defines a new species of particles in the domain (e.g. `reflected_electrons`), the current report will probably need to be changed. I'm not completely certain of the diagnostics structure that you're using, but the `_extract_current()` function probably only reads in particle information from one species. This would need to be augmented to include the reflected particles as well, if we were to compute the total current within the domain.

    Regarding the Particle Trace report, I imagine that there won't be significant issues, because you can simply ignore the additional species designations. My only recommendation is that you change the labels on the report (which you may have already done). Originally we used the term "Reflected" to refer to particles that didn't cross the gap owing to the existence of a retarding potential. This is now a confusing descriptor. "Reflected" should refer only to particles that strike an object and are reflected using our new class. You could use the term "Rejected" (or something similar) instead for the particles that don't make the crossing on their own.

2. The object does not have to be a conductor in the physical sense. Dielectric objects or semiconductors could similarly reflect particles. I can't speak to the subtle differences in the physics here. If it's a real pain to support both, then it's OK to stick to just conducting objects.

3. No. The reflection process is generally independent of any external biasing, since the scale length variation in the "reflecting potential" is orders of magnitude smaller than the scale length variation in the applied potential. A conductor held at zero potential will reflect particles all the same.

4. Reflection is supported (and has been tested) in 3D geometry. Nothing in the implementation should preclude it from working in 2D, but I haven't tested it yet. I recommend that you support it for both, since that's probably easiest. If we run into issues after some testing, we can correct them or remove it in 2D.

5. Yes, both the cathode and anode can have reflection. In general, reflection at the anode is the single most important reflection process to include. You can probably ignore reflection at the cathode for now, because it's statistically much less likely in most instances.

6. I recommend limiting the GUI to diffuse reflection only (set 0 probably for specular reflection), and suggest fixing the `refscheme` to be `uniform` for now. The use of specular reflection in these simulations is largely unsupported by data, and furthermore it's probably strongly affected by surface defects which we do not capture.

**Comment by mkeilman** [comment]
Another question: by default the particle boundary condition at the anode and cathode is "absorb".  Does that conflict with reflections?  I see there is a "reflect" condition as well...

**Comment by ncook882** [comment]
Under most scenarios, the particle boundary conditions at the left and right edge of the domain (`pbound0` and `pboundnz`) should not come into play, because the particles should be removed by the scrapers associated with the cathode and anode conductor slices.

That said, when running in parallel, there are times wherein the anode scraper doesn't work properly, and some particles will first be scraped by the domain boundaries prior to striking the anode. Because of this, it's safer to use the "absorb" boundary condition. Introducing "reflect" boundary conditions would result in a specular-like reflection of the particles and would definitely skew the physics of the simulation.

Regardless of these conditions, the `pboundxy` should remain periodic.
